src/schema.ts

```diff
@@ -0,0 +1,385 @@
+ * - [[State]]
+ * - [[DropEvent]] (and subtypes/interfaces)
+ * -------
+ * - [[Performance]]
+ * - [[PerformanceMetadata]]
+ * - Events
+ * -------
+ * Ordered by Machine, Vibraphone, Bass,
+ * HiHat machine, HiHat:
+ * - sub-States
+ * - sub-Events
+ * @packageDocumentation
+ */
+import { Note } from "./note_names";
+
+/**
+ * Represents all possible drums that can be played. Used for specifying which drum
+ * to play in a [[DrumDropEvent]].
+ */
+export type DrumType = "bassdrum" | "hihat" | "snare";
+/**
+ * Represents a channel that can be muted. Used as keys in the [[MachineState.mute]] object.
+ *
+ * ## Example
+ * ```typescript
+ * // specifies that the bassdrum and vibraphone are muted
+ * const state: MachineState = {
+ *   mute: {
+ *     bassdrum: true,
+ *     vibraphone: true,
+ *   },
+ *   ...
+ * };
+ */
+export type Channel = DrumType | "vibraphone" | "bass";
+
+/**
+ * Represents a program on the MMX's programming wheel. Used to store the programmed part of a performance.
+ */
+export interface Program {
+	/**
+	 * Information about this program. It can be distinct from a performance's metadata.
+	 */
+	metadata: ProgramMetadata;
+	/**
+	 * This is the "working" state of the machine, outside of a performance.
+	 * When a user is modifying the program, this is how we store what settings they use.
+	 * Whenever the user saves, this field gets updated, and whether they load, this gets read.
+	 * It has almost no relevance when editing a performance since the performance events will
+	 * cause it to be updated dynamically.
+	 *
+	 * ## Example
+	 * ```typescript
+	 * let prog: Program;
+	 * // invert the muted state of the snare
+	 * prog.state.machine.mute.snare = !prog.state.machine.mute.snare;
+	 * ```
+	 */
+	state: State;
+	/**
+	 * All the [[TickedDropEvent]]s that occur during this program. They *must* be in ascending order by tick.
+	 */
+	dropEvents: TickedDropEvent[];
+}
+
+/** Represents metadata about a program. Used to indicate how this program should be played. */
+export interface ProgramMetadata {
+	/**
+	 * The title of this program. It can be distinct from the title of the performance.
+	 */
+	title: string;
+	/**
+	 * The author of this program. It can be distinct from the title of the performance.
+	 */
+	author: string;
+	/**
+	 * Pulses per quarter.
+	 * This basically represents how many distinct pieces a quarter note is divisible by.
+	 * The higher the PPQ, the more flexibility when it comes to placing notes.
+	 */
+	readonly tpq: 240;
+	/** Version of VMMX in which the current program was made */
+	readonly version: string;
+	/** Total ticks on the "programming wheel" 240 PPQ \* 4 beats/measure \* 16 bars on wheel */
+	readonly length: 61440;
+	/** The amount of procrastination that occurred during the making of this program. */
+	procrastination?: number;
+}
+
+/** Represents the machine's state. Used to specify both the start state and running state of the machine. */
+export interface State {
```

Comment:
Maybe some comments for each state
```
machine: MachineState; // channel mutes, bpm, flywheel connected
vibraphone: VibraphoneState; // vibrato enabled, vibrato speed (also dampening when implemented)
bass: BassState; // Capo positions and tuning
hihatMachine: HihatMachineState; // setting (whole note / offbeat / 16ths / triplet) (not specified in schema yet)
hihat: HihatState; // closed or open
```

Reply:
If people want to know what properties the different states have, they can simply look it up in the specific State definition.